### ISSUE TITLE
Why are the uint and int functions opposite in terms of default bounds? 

### ISSUE THREAD

**Issue by Chris3606** [pomma89/Troschuetz.Random#5]
Looking at IGenerator, I see that `Next()` produces an int between 0, and `int.MaxValue` _exclusive_, and `NextInclusiveMaxValue()` is between 0 and `int.MaxValue` _inclusive_.  Looking at the `uint` functions, I see the opposite: `NextUInt()` produces a value between 0 and `uint.MaxValue` _inclusive_, whereas `NextUIntExclusiveMaxValue()` produces a value between 0 and `uint.MaxValue` _exclusive_.

Why does `NextUInt()` default to an _inclusive_ bound, whereas `Next()` defaults to an _exclusive_ bound?  At first glance it seems a little unintuitive -- I would expect that both would default to the same bounds behavior, if there is no reason to do otherwise.

**Comment by pomma89** [comment]
Hi @Chris3606,

If I understood correctly, you are asking two questions:

1. Isn't it confusing to have `Next` and `NextUInt` behaving differently?
2. Why was it implemented like that?

For the first question, the answer is yes, it was a bad choice to have two methods with similar names behaving differently. `Next` followed [.NET `Random` class](https://docs.microsoft.com/it-it/dotnet/api/system.random.next?view=netframework-4.7#System_Random_Next), while `NextUInt` implemented a different logic and the reasoning is part of second question.

In fact, implementing `NextUInt` differently seemed a good choice from the performance point of view, because generators naturally produce 64 random bits and returning a 32 bit unsigned integer is just a matter of taking half bits. If an upper bound has to respected, then the implementation has to be slower in order to check the bound, like this:

```cs
public uint NextUIntExclusiveMaxValue()
{
    uint result;
    do result = NextUInt(); while (result == uint.MaxValue);

    // Postconditions
    Debug.Assert(result < uint.MaxValue);
    return result;
}
```

As I said, I understand that this choice is unintuitive. Changing the implementation is not difficult, of course, but the interface would be broken in a subtle way (`NextUInt` would stop returning `MaxValue`, which might be a problem for someone). In any case, `NextUIntExclusiveMaxValue` would continue to exist and `NextUIntInclusiveMaxValue` would be added.

I need to think a bit about this and I'm glad to receive some feedback.

Thanks.

**Comment by pomma89** [comment]
Just released v4.3.0:

* `NextUInt` uses the same logic as `Next`
* Deprecated `NextUIntExcludingMaxValue`
* Added `NextUIntIncludingMaxValue`

Thanks.
